### PR TITLE
Multi-process platform mode.

### PR DESCRIPTION
**Pull request by sas**

### CODE REVIEW

**Comment by fjricci** [comment]
Can you try uploading just the first 2 commits and see if the flakiness goes away? It should, but if it doesn't, it helps tell us what the issue is.

File: Sources/main.cpp [hunk 1]
```diff
@@ -83,19 +84,43 @@ CreateSocket(std::string const &host, std::string const &port, bool reverse) {
...
-    session.setDelegate(&impl);
-    session.create(client.get());
+  std::unique_ptr<Socket> serverSocket = CreateSocket(host, port, false);
 
-    while (session.receive(/*cooked=*/false))
-      continue;
+  do {
+    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
+    auto platformClient =
+        std::make_shared<PlatformClient>(std::move(clientSocket));
```
Comment (fjricci) [inline Sources/main.cpp]: unique instead of shared?

File: Sources/main.cpp [hunk 2]
```diff
@@ -83,19 +84,35 @@ CreateSocket(std::string const &host, std::string const &port, bool reverse) {
...
-    // Platform mode implies that we are talking to an LLDB remote.
-    Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
-    session.setDelegate(&impl);
-    session.create(client.get());
-
-    while (session.receive(/*cooked=*/false))
-      continue;
+    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
+    auto platformClient =
+        ds2::make_unique<PlatformClient>(std::move(clientSocket));
```
Comment (strager) [inline Sources/main.cpp]: Why not create this inside the thread? Error handling?

Why use a unique_ptr? Is one of the members of PlatformClient not movable?

File: Sources/main.cpp [hunk 3]
```diff
@@ -83,19 +84,35 @@ CreateSocket(std::string const &host, std::string const &port, bool reverse) {
...
+
+    std::thread thread(
+        [](std::unique_ptr<PlatformClient> client) {
+          while (client->session.receive(/*cooked=*/false)) {
+            continue;
+          }
+        },
+        std::move(platformClient));
+
+    thread.detach();
```
Comment (strager) [inline Sources/main.cpp]: Why are multiple threads preferred over handling multiple connections on a single thread?

Comment (sas) [inline Sources/main.cpp]: Because it is easier to just create a new thread every time a client
connects. Doing otherwise would require polling and such.

On Monday, July 11, 2016, strager <notifications@github.com> wrote:

> In Sources/main.cpp
> <https://github.com/facebook/ds2/pull/403#discussion_r70359489>:
>
> > -
> > -    while (session.receive(/*cooked=*/false))
> > -      continue;
> > +    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
> > +    auto platformClient =
> > +        ds2::make_unique<PlatformClient>(std::move(clientSocket));
> > +
> > +    std::thread thread(
> > +        [](std::unique_ptr<PlatformClient> client) {
> > +          while (client->session.receive(/*cooked=*/false)) {
> > +            continue;
> > +          }
> > +        },
> > +        std::move(platformClient));
> > +
> > +    thread.detach();
>
> Why are multiple threads preferred over handling multiple connections on a
> single thread?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/ds2/pull/403/files/e6d0f35a1ebb85fe8c6e1d0e91b19a4d9aee6752..024d6577131834ae01bc46af0838d244b7588467#r70359489>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADHbJyI_Qr0G9mPuiDGpRKG2jV7phFEks5qUtwGgaJpZM4JIJxl>
> .
>


-- 
Stephane Sezer

File: Sources/main.cpp (hunk 2 above)
Comment (sas) [inline Sources/main.cpp]: Create what? The unique_ptr or the PlatformClient?

The unique_ptr is create here so we can get rid of standard pointer
operations, which means error handling is in general easier, and leaks are
more rare.

I am using a pointer for this so I don't have to copy the whole
PlatformClient structure, which can get pretty big.

On Monday, July 11, 2016, strager <notifications@github.com> wrote:

> In Sources/main.cpp
> <https://github.com/facebook/ds2/pull/403#discussion_r70359359>:
>
> >
> >    do {
> > -    std::unique_ptr<Socket> client = server->accept();
> > -    // Platform mode implies that we are talking to an LLDB remote.
> > -    Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
> > -    session.setDelegate(&impl);
> > -    session.create(client.get());
> > -
> > -    while (session.receive(/*cooked=*/false))
> > -      continue;
> > +    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
> > +    auto platformClient =
> > +        ds2::make_unique<PlatformClient>(std::move(clientSocket));
>
> Why not create this inside the thread? Error handling?
>
> Why use a unique_ptr? Is one of the members of PlatformClient not movable?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/ds2/pull/403/files/e6d0f35a1ebb85fe8c6e1d0e91b19a4d9aee6752..024d6577131834ae01bc46af0838d244b7588467#r70359359>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADHbIASlTG9kUWBrjPQOkD2QVA4hvgcks5qUtukgaJpZM4JIJxl>
> .
>


-- 
Stephane Sezer

Comment (strager) [inline Sources/main.cpp]: Why not:

    std::unique_ptr<Socket> clientSocket = serverSocket->accept();

    std::thread thread(
        [&impl](std::unique_ptr<Socket> clientSocket) {
          // Platform mode implies that we are talking to an LLDB remote.
          Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
          session.setDelegate(&impl);
          session.create(clientSocket.get());
          while (session.receive(/*cooked=*/false)) {
            continue;
          }
        },
        std::move(clientSocket));

Comment (sas) [inline Sources/main.cpp]: Where does impl come from?

On Monday, July 11, 2016, strager <notifications@github.com> wrote:

> In Sources/main.cpp
> <https://github.com/facebook/ds2/pull/403#discussion_r70362758>:
>
> >
> >    do {
> > -    std::unique_ptr<Socket> client = server->accept();
> > -    // Platform mode implies that we are talking to an LLDB remote.
> > -    Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
> > -    session.setDelegate(&impl);
> > -    session.create(client.get());
> > -
> > -    while (session.receive(/*cooked=*/false))
> > -      continue;
> > +    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
> > +    auto platformClient =
> > +        ds2::make_unique<PlatformClient>(std::move(clientSocket));
>
> Why not:
>
> std::unique_ptr<Socket> clientSocket = serverSocket->accept();
>
> std::thread thread(
>     [&impl](std::unique_ptr<Socket> clientSocket) {
>       // Platform mode implies that we are talking to an LLDB remote.
>       Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
>       session.setDelegate(&impl);
>       session.create(clientSocket.get());
>       while (session.receive(/*cooked=*/false)) {
>         continue;
>       }
>     },
>     std::move(clientSocket));
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/ds2/pull/403/files/e6d0f35a1ebb85fe8c6e1d0e91b19a4d9aee6752..024d6577131834ae01bc46af0838d244b7588467#r70362758>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADHbIi62MC2rvlafBoDPkv_rOzp25atks5qUuTxgaJpZM4JIJxl>
> .
>


-- 
Stephane Sezer

Comment (strager) [inline Sources/main.cpp]: Put outside the loop:

  PlatformSessionImpl impl;

Comment (sas) [inline Sources/main.cpp]: We have one impl per client, we can't do that.

On Mon, Jul 11, 2016 at 6:31 PM, strager <notifications@github.com> wrote:

> In Sources/main.cpp
> <https://github.com/facebook/ds2/pull/403#discussion_r70365608>:
>
> >
> >    do {
> > -    std::unique_ptr<Socket> client = server->accept();
> > -    // Platform mode implies that we are talking to an LLDB remote.
> > -    Session session(ds2::GDBRemote::kCompatibilityModeLLDB);
> > -    session.setDelegate(&impl);
> > -    session.create(client.get());
> > -
> > -    while (session.receive(/*cooked=*/false))
> > -      continue;
> > +    std::unique_ptr<Socket> clientSocket = serverSocket->accept();
> > +    auto platformClient =
> > +        ds2::make_unique<PlatformClient>(std::move(clientSocket));
>
> Put outside the loop:
>
> PlatformSessionImpl impl;
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/ds2/pull/403/files/e6d0f35a1ebb85fe8c6e1d0e91b19a4d9aee6752..024d6577131834ae01bc46af0838d244b7588467#r70365608>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADHbJ_UZUA8uJENCqvUkH74MzHfAi70ks5qUu5wgaJpZM4JIJxl>
> .
>



-- 
Stephane Sezer

Comment (strager) [inline Sources/main.cpp]: Fine, create it in the thread function instead. =]

**Comment by sas** [comment]
This exposes some bugs in platform mode. Can't checkin for now.
